contracts/tge/HATAirdropFactory.sol

```diff
@@ -0,0 +1,115 @@
+// SPDX-License-Identifier: MIT
+// Disclaimer https://github.com/hats-finance/hats-contracts/blob/main/DISCLAIMER.md
+
+pragma solidity 0.8.16;
+
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/proxy/Clones.sol";
+import "./HATAirdrop.sol";
+
+contract HATAirdropFactory is Ownable {
+    using SafeERC20Upgradeable for IERC20Upgradeable;
+
+    address public implementation;
+
+    event ImplementationUpdated(address indexed _newImplementation);
+    event TokensWithdrawn(address indexed _owner, uint256 _amount);
+    event HATAirdropCreated(
+        address indexed _hatAirdrop,
+        string _merkleTreeIPFSRef,
+        bytes32 _root,
+        uint256 _startTime,
+        uint256 _deadline,
+        uint256 _lockEndTime,
+        uint256 _periods,
+        uint256 _totalAmount,
+        IERC20Upgradeable _token,
+        ITokenLockFactory _tokenLockFactory
+    );
+
+    constructor (address _implementation) {
+        implementation = _implementation;
+    }
+
+    function updateImplementation(address _newImplementation) external onlyOwner {
```

Comment:
this is nice but it is not as flexible as it looks, bc any implementation contract needs to use the same signature for the initialization function. If we want the flexibility to create new airdrop contracts, we proably should pass the arguments encoded in a single parameter (but also that has it's drawbacks in terms of ux)

Reply:
Ok yeah it'd be more flexible